#213 Fix recalculations when deleting a repeat group

mdudzinski opened the pull request:
Closes #213

#### What has been done to verify that this works as intended?
So far only manual testing. Unit tests are in progress.

#### Why is this the best possible solution? Were any other approaches considered?
It seems to be most 'natural' to me. However there were other approaches considered. More details in the conversation under #213.
I think a really slow-down may be acceptable as `Safe2014DagImpl` seems to focous on being bugs-free. There are other implementations - like `Fast2014DagImpl` - which seem to be focused on the performance. Even at the expense of some bugs or not being as accurate as the safe implementations.

#### Are there any risks to merging this code? If so, what are they?
Well, when deleting a repeat, the default evaluation logic (`Safe2014DagImpl`) goes over all remaining repeats and fires all recalculations that were depended on the remaining repeat groups (their position) or their children. It might have a performance impact if there are a lot of repeats with depended calculations. I haven't tested it yet though.

mdudzinski commented:
@ln 
This one can be reviewed and/or tested.
I think that I figured out a better way for testing triggerables evaluation. It doesn't use API classes like `FormEntryController` and `FormEntryModel`. Instead, the test parsse a form with main instance that has test data and calls `Safe2014DagImpl#deleteRepeatGroup` in an isolation.

mdudzinski commented:
@ln @dcbriccetti This one also includes a proposition for managing test forms. We'll keep adding them to test various new or existing functionalities and having all of them in a same directory may become cumbersome at some point. So I propose keeping test forms in the same packages (inside resources directory) as the test classes. 
The only drawback I can see is a need to put the full path to the test form when writing a test. But I think the test form parser could be improved so it could infer the path to a given form basing on the package name of the test class.

lognaturel commented:
@mdudzinski and I discussed in Slack. This will have performance implications on instances that have lots of repeats. Would be good to get a sense of how significant the impact is.

dcbriccetti commented:
>Would be good to get a sense of how significant the impact is.

Shall I wait for that before reviewing?

lognaturel commented:
I think a quick pass with first impressions would be helpful.

dcbriccetti commented on test/org/javarosa/test/utils/FormParserHelper.java:
```diff
@@ -0,0 +1,47 @@
+package org.javarosa.test.utils;
```
I like this change, but is there a reason you moved the classes out of the xform.parse package? I think they belong there, rather than in a new “utils” package.

mdudzinski replied:
@dcbriccetti Since this is code is now used by tests outside of the`xform.parse` package, I wanted to put it in a common place, not related to any particular feature of the production code (like xform parsing).

dcbriccetti replied:
I’m not sure I understand. If the classes are related to parsing, it makes sense to have them in a `parse` package, even if they are used from classes outside the package.

mdudzinski replied:
@dcbriccetti Hmm I just treated this class as a helper class and wanted to keep it outside of packages related to the actual production code. On second thought, it actually tests the form parsing codebase so it *is* related to the production code. I'll move it back to the `xform.parse` package. Thanks!

mdudzinski replied:
@dcbriccetti Done!

dcbriccetti commented on test/org/javarosa/test/utils/ResourcePathHelper.java:
```diff
@@ -0,0 +1,11 @@
+
+public class ResourcePathHelper {
+    /** Makes a Path for a resource file */
+    public static Path r(String filename) {
```
I really like moving the forms within the package under resources, but why not go “all the way to town”?

```
/** Makes a Path for a resource file */
public static Path r(String filename) {
    return Paths.get("resources", filename);
}

/**
 * Makes a Path for a resource file in the package identified by <code>classForPackage</code>
 * @param classForPackage the class whose path is to be used to locate the file in the resources directory
 * @param filename the name of the file in the subdirectory of the resources directory
 * @return a Path
 */
public static Path r(Class<?> classForPackage, String filename) {
    return Paths.get("resources", classForPackage.getPackage().getName().replace('.', '/'), filename);
}
```

Call like this:
`r(getClass(), "repeat-group-with-position-calculation.xml")`

dcbriccetti replied:
I also like `r(this, ...)` because it’s shorter, but some might be bothered (myself included) that far more than needed was passed as an argument.

mdudzinski replied:
@dcbriccetti Actually I thought about this but wanted to use `Thread.currentThread().getStackTrace()` to automatically infer the package of the calling class. I'm fine with both approaches.

dcbriccetti replied:
>getStackTrace

Oh, that’s the best! It hadn’t occurred to me.

mdudzinski replied:
@dcbriccetti Done 😃

dcbriccetti commented on test/org/javarosa/core/model/Safe2014DagImplTest.java:
```diff
@@ -0,0 +1,104 @@
+        assertThat(repeats.get(2).getChildAt(0).getValue().getDisplayText(), equalTo("3"));
+        assertThat(repeats.get(3).getChildAt(0).getValue().getDisplayText(), equalTo("4"));
+
+        // check that calculations have not been triggered for the repeat group prior to the deleted one
```
Would you consider this, to improve readability and maintainability?

```
// check that calculations have not been triggered for the repeat group prior to the deleted one
final String[] expectedMessages = {
        "Processing 'Recalculate' for no [2_1] (2.0)",
        "Processing 'Deleted: houseM [2]: 1 triggerables were fired.' for ",
        "Processing 'Deleted: no [2_1]: 1 triggerables were fired.' for ",
        "Processing 'Recalculate' for no [3_1] (3.0)",
        "Processing 'Deleted: houseM [3]: 1 triggerables were fired.' for ",
        "Processing 'Deleted: no [3_1]: 1 triggerables were fired.' for ",
        "Processing 'Recalculate' for no [4_1] (4.0)",
        "Processing 'Deleted: houseM [4]: 1 triggerables were fired.' for ",
        "Processing 'Deleted: no [4_1]: 1 triggerables were fired.' for "
};
int messageIndex = 0;
for (String expectedMessage : expectedMessages) {
    assertThat(dagEvents.get(messageIndex++).getDisplayMessage(), equalTo(expectedMessage));
}
```

mdudzinski replied:
@dcbriccetti I love this! Will use this approach.

dcbriccetti commented on test/org/javarosa/core/model/Safe2014DagImplTest.java:
```diff
@@ -0,0 +1,104 @@
+        final TreeReference dummyDeletedRef = null;
+
+        // When
+        dag.deleteRepeatGroup(mainInstance, formDef.getEvaluationContext(), dummyDeletedRef, parentElement, deletedElement);
```
I’m wondering if it would be safer, more realistic, and/or simpler to call formdef.deleteRepeat(FormIndex index) rather than getting the dag and calling deleteRepeatGroup directly.

mdudzinski replied:
@dcbriccetti Well, I mostly wanted to test dag in isolation but it should be possible to test it via `formdef.deleteRepeat(FormIndex index)`. 
Since I'd like to avoid using API classes like `FormEntryModel` and/or `FormEntryController`, the `FormIndex index` would need to be constructed from scratch. After my time spent on #218 (where I was analyzing FormIndex inc/dec codebase) , it shouldn't be a problem now.

mdudzinski replied:
@dcbriccetti I followed your suggestion and now the test looks way better. Thanks!

dcbriccetti commented on test/org/javarosa/core/model/Safe2014DagImplTest.java:
```diff
@@ -0,0 +1,104 @@
+
+        formDef.setEventNotifier(eventNotifier);
+        IDag dag = getIDagImplUnderTest(formDef);
+        assertThat(dag, instanceOf(Safe2014DagImpl.class));
```
getIDagImplUnderTest casts to this, so you’ll crash before failing this assertion.

mdudzinski replied:
@dcbriccetti Nice catch! Thanks

mdudzinski replied:
@dcbriccetti Fixed

dcbriccetti commented:
Hey, this is complex stuff! Hats off to you, @mdudzinski. I made a few review comments, and I’ll continue later today or tomorrow.

mdudzinski commented:
@lognaturel 
I tested the performance (on an AVD) impact on the following form:
[delete-repeat-recalculate.xml.txt](https://github.com/opendatakit/javarosa/files/1495912/delete-repeat-recalculate.xml.txt)
It has 100 repeats with the following calculations:
```
            <bind calculate="position(..)" nodeset="/data/household/index" type="int"/>
            <bind calculate="log(pow(position(..), 5))" nodeset="/data/household/somecalculation" type="decimal"/>
```
and there was no performance impact. Or at least it wasn't noticeable. 

To get more reliable test results, I asked @mmarciniak90 to test this form on an actual device (Thanks Marzena!).
She tested this on Android 4.4 and 4.1 and confirmed there were no noticeable performance impact as well.

I'm now working on fixing/implementing suggestions and concerns from the review. It should be available for the review again soon.

mdudzinski commented:
@dcbriccetti @lognaturel I updated this PR, please review it when you get a chance. Thanks!

lognaturel commented on src/org/javarosa/core/model/Safe2014DagImpl.java:
```diff
@@ -61,13 +61,20 @@ public void deleteRepeatGroup(FormInstance mainInstance,
-      evaluateChildrenTriggerables(mainInstance, evalContext,
-            deletedElement, false, alreadyEvaluated);
+      // After a repeat group has been deleted, the following repeat groups position has changed.
+      // Evaluate triggerables which depend on the position of the following repeats or on their children.
```
This makes it sound like it only evaluates certain triggerables. Maybe it should say something more like "evaluate triggerables for all following repeats and their children?" Let me know if I'm just misunderstanding the wording.

mdudzinski replied:
@lognaturel  At the first glance it looks like it evaluates only directly related triggerables like `position(..)` but `Safe2014DagImpl` is super smart and will also fire all indirectly related ones. For example:
`<bind calculate="concat(/rgwp/houseM/name, /rgwp/houseM/no)" nodeset="/rgwp/houseM/namecalc" type="string"/>`
This one uses `/rgwp/houseM/no` value which explicitly depends on the position. The dag implementation will fire the concatenation calculation after the position calculation for the `/rgwp/houseM/no` node. 

On the other hand, the following 
`<bind calculate="concat(/rgwp/houseM/name, 'some string value')" nodeset="/rgwp/houseM/namecalc" type="string"/>`
won't be fired because the concatenation doesn't rely at all on the position within the parent. 
This is pretty amazing I think.

Anyway, both "evaluate triggerables for all following repeats and their children" and " evaluate triggerables which depend on the position of the following repeats or on their children" may be confusing. Perhaps we should document the above with a new unit test scenario?

lognaturel replied:
That's cool! So we ended up with your "ideal" solution after all.

Adding a test and doing your best to clarify the comment both sound good. Perhaps in the comment you could link to the relevant tests.

mdudzinski replied:
@lognaturel Ayay I must've made a mistake yesterday while debugging this! It's smart but not as smart as I thought. It doesn't re-evaluate same triggerable when a repeat is deleted but it will re-evaluate any other triggerable inside of the repeat. 
So all I said for
`<bind calculate="concat(/rgwp/houseM/name, /rgwp/houseM/no)" nodeset="/rgwp/houseM/namecalc" type="string"/>` 
is correct but it is not for
`<bind calculate="concat(/rgwp/houseM/name, 'some string value')" nodeset="/rgwp/houseM/namecalc" type="string"/>`.

And I know why this happens. DAG goes over each child of each repeat and looks for triggerables that are dependent on the child reference. In this scenario, it finds `calculate="concat(/rgwp/houseM/name, 'some string value')" `. This should be improved because this triggerable depends on `/rgwp/houseM/name` which doesn't depends on a calculation related to the parent reference - `/rgwp/houseM`. 
I can think about fixing it now or raise an issue for this with some information I already have.

Sorry for the confusion. It came out when I created a test for this. It shows how desperate JavaRosa needs tests.
So no ideal solution for now 😞

mdudzinski replied:
@lognaturel if fixing the above behavior isn't trivial, I can test some way more complex calculations in a repeat and see if there is still no impact on the performance.

dcbriccetti commented on test/org/javarosa/test/utils/ResourcePathHelper.java:
```diff
@@ -0,0 +1,35 @@
+     * then this method will return "org/javarosa/core/model"
+     * @return Caller test class package as a path
+     */
+    private static String inferResourceFileParentPath() {
```
I love this, and would not have thought of it. How about making it more resilient, though, by removing the assumption that the caller will always be exactly two level up. Something like this, perhaps?

```
    private static String inferResourceFileParentPath() {
        final StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        int callerStackIndex = 2; // 0 is getStackTrace, 1 is this method, 2 is immediate caller
        while (callerStackIndex < stackTrace.length &&
                stackTrace[callerStackIndex].getClassName().equals(ResourcePathHelper.class.getName())) {
            ++callerStackIndex;
        }
        final StackTraceElement callingClass = stackTrace[callerStackIndex];
        final String callerClassName = callingClass.getClassName();
        return callerClassName
                .substring(0, callerClassName.lastIndexOf(".")) // strip the class name
                .replaceAll("\\.", "\\" + File.separator); // change all '.' to '/' ('\' on Windows)
    }
```

mdudzinski replied:
@dcbriccetti Why not! I'll take this, thanks!

mdudzinski replied:
@dcbriccetti done!

dcbriccetti commented on test/org/javarosa/test/utils/ResourcePathHelper.java:
```diff
@@ -0,0 +1,35 @@
+        final String callerPackage = callingClass.getClassName();
+        return callerPackage
+                .substring(0, callerPackage.lastIndexOf(".")) // strip the class name
+                .replaceAll("\\.", "\\" + File.separator); // change all '.' to '/' ('\' on Windows)
```
`"\\" + File.separator` gives me `"\/"` (on Mac). Is that better than `"/"`? The second argument to replaceAll is a simple string. But, why not avoid regexes altogether and use `replace(".", File.separator)`?

mdudzinski replied:
@dcbriccetti it just worked on my mac so I didn't check the output. I'll take a look at your suggestion. Thanks!

mdudzinski replied:
@dcbriccetti But anyway I forgot about `replace` method which doesn't use regexes. I always try to avoid them whenever possible 😃 
As  Jamie Zawinski said:
>Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems.

mdudzinski replied:
@dcbriccetti done!

dcbriccetti commented:
>I tested the performance (on an AVD) impact on the following form: delete-repeat-recalculate.xml.txt

Might it be useful to make a test that does many deletes from that form, so we can time it? Or are we confident that performance is acceptable? Might there be actual forms with many thousands of elements, such as there is with the Nigeria Wards form?

mdudzinski commented:
> Might it be useful to make a test that does many deletes from that form, so we can time it? Or are we confident that performance is acceptable? Might there be actual forms with many thousands of elements, such as there is with the Nigeria Wards form?

There is no point in timing many deletes because you can't do this in batch. What matters is a time of a single deletion when there is a lot of following repeats with calculations to be recomputed. 
And it's important that not every calculation will be recomputed, only those that actually depend on the parent position. More details can be found [here](https://github.com/opendatakit/javarosa/pull/219#discussion_r153533480).

Of course it's possible to create a form with way more complex computation depending on the parent position than we tested [here](https://github.com/opendatakit/javarosa/pull/219#issuecomment-346336293) and at some point a user will start seeing the performance impact but not sure if that would be a real life production scenario.

mdudzinski commented:
Updated PR according to the latest suggestions from the code review but I'm still about to add new tests and improve a bit the comments around the changed method.

lognaturel commented:
Great that the tests are helping uncover interesting things! I think creating an issue describing the triggerables that are reevaluated but that shouldn't need to be is a good idea. Seems adding in a timing test would be valuable as well. I think what @dcbriccetti is suggesting is to simulate multiple deletes to get an average timing value that is representative of reality. Would be good to try up to 1000 repeats.

mdudzinski commented:
@lognaturel Sure, I can add a timing test. Any idea how much complicated calculations should be in the test form?

Now I'm investigating calling ` evaluateChildrenTriggerables(mainInstance, evalContext,          repeatGroup, false, alreadyEvaluated)` for each following repeat. This triggers unnecessary re-calculations for some of my test forms but I'm wondering if there is a production scenario where this re-evaluation is *not* unnecessary.

lognaturel commented:
@mdudzinski I don't think the calculations need to be very complicated. My guess is that position is generally used only for display and not as part of larger expressions. Are you thinking about the complexity of other child/sibling calculations as well? Something like a few integer fields and a calculation of their sum seems like it would be representative.

mdudzinski commented:
@lognaturel Okay so I did a tweak to the fix and now only children of the actually deleted repeat have their triggerables evaluated. Turned out it's unnecessary to evaluate children triggs for all the following repeats. I added bunch of tests to cover this and other scenarios. In the last commit - still marked as wip - I added a timing test and honestly I don't like the results... The test form is `delete-repeat-group-with-calculations-timing-test.xml` and there are couple of bindings commented to not stress the CI 😛 

So where there are 1000 repeats and the following bindings:
```
<bind calculate="position(..)" nodeset="/rgwp/houseM/current_position" type="int"/>
<bind calculate="pow(/rgwp/houseM/current_position, 3)" nodeset="/rgwp/houseM/int1" type="int"/>
<bind calculate="log10(/rgwp/houseM/int1)" nodeset="/rgwp/houseM/double1" type="double"/>
<bind calculate="round(/rgwp/houseM/double1)" nodeset="/rgwp/houseM/int2" type="int"/>
<bind calculate="sum(/rgwp/houseM/current_position)" nodeset="/rgwp/summary" type="double"/>
```
deleting one repeat takes about 4.25 seconds. Since the test deletes all the repeats starting from the first one and the fix evaluates triggerables only for the following groups, this time should be decreasing with each iteration (because of less repeat group to iterate through).

Interesting is that calculations inside of the repeat (pow, log and round) have very little impact on this time and this test run against:
```
<bind calculate="position(..)" nodeset="/rgwp/houseM/current_position" type="int"/>
<bind calculate="sum(/rgwp/houseM/current_position)" nodeset="/rgwp/summary" type="double"/>
```
shows that deletion of a single repeat takes about 3.9 ~ 4 seconds. 

So the sum function is a bottleneck here. When you take it out and leave the remaining calculations, everything works quite smoothly (100 - 200 millis for a single deletion at the beginning of the loop).

When you reduce the amount of repeats to 500 and leave the sum, then deleting a repeat group at the beginning of the loop takes about 0.4 - 0.5 seconds. At the end it takes less than 0.001 of a second.

mdudzinski commented:
This also means that if you deleting a 800th repeat group will be way faster then deleting a second one.

dcbriccetti commented on test/org/javarosa/core/model/Safe2014DagImplTest.java:
```diff
@@ -0,0 +1,396 @@
+            String tookFormatted = LocalTime.fromMillisOfDay(took).toString();
+            System.out.println("Deletion of repeat #" + i + " took " + tookFormatted);
+        }
+
```
Just a thought. If you do something like this:
```
System.out.println("Repeat Group\nDeletion Time (ms)");
for (int i = 0; i < numberOfRepeats; i++) {
    long currentIterationStart = System.nanoTime();
    formDef.deleteRepeat(firstRepeatIndex);
    double tookMs = (System.nanoTime() - currentIterationStart) / 1000000D;
    System.out.printf("%d\t%.3f\n", i, tookMs);
}
```
Then we can do neat things like this:

<img width="697" alt="deletion times" src="https://user-images.githubusercontent.com/55099/33785488-185580f0-dc1a-11e7-8170-87e91bf0f7cd.png">

mdudzinski replied:
@dcbriccetti Thanks, I'll fix the output to match this format.

@lognaturel 
If the deletion times are acceptable then I'd like to clean up the test so it can be merged. The number of iterations should be decreased to not slow down build times - what would say about lowering it to 200 elements? Deleting all repeats takes about 3.5 seconds in that case.

mdudzinski replied:
@dcbriccetti @lognaturel I pushed the clean-up timing test. It uses format suggested by @dcbriccetti and tests deletion of 200 repeats.

dcbriccetti replied:
What we did for the timing “tests” in XFormParserTest was make them run very fast (a fraction of a second), but have a comment with instructions for anyone wanting to make them run longer, for actual timing purposes:
`        for (double powerOfTen = 3; powerOfTen <= 4.0; powerOfTen += 0.1) {  // Raise this upper limit to really measure
`

mdudzinski replied:
@dcbriccetti Right, thanks! I added a comment and left 200 repeats. It runs fast and it's still quite a lot.

lognaturel commented:
Cleaning things up and getting it ready for merging sounds great. I think we can get this into the next Collect release — code freeze Wednesday and release on Sunday or so.

lognaturel commented:
Thanks for approaching this in such a thoughtful way, @mdudzinski! The test cases are really good and I can't think of others. These are really informative tests.

I understand that `sum` takes a lot less time when there are fewer repeats overall -- when you've deleted 400 out of 500 for example.

> This also means that if you deleting a 800th repeat group will be way faster then deleting a second one.

This makes it sound like if I have a total of 1000 repeat and delete one in either the 800th or 2nd position the time to compute the sum will be different. Is that really the case? If so, how? I wouldn't expect there to be a shortcut for that one -- seems it would just have to re-compute the whole thing. Similarly, I would expect that adding a new repeat (say, the 1001st) would lead to the `sum` outside the repeat being re-computed and taking just as much time as a delete in that case. Is that right?

Other than that, this looks ready to merge. We'll ask people to try this out in the beta. And we'll just be ready to revert or adjust in case of any problems in production.

lognaturel commented:
Well, I had to satisfy my curiosity -- 
```
        final FormIndex indexToDelete = new FormIndex(0, 3, mainInstance.getRoot().getChildAt(4).getRef());
        long currentIterationStart = System.nanoTime();
        formDef.deleteRepeat(indexToDelete);
        double tookMs = (System.nanoTime() - currentIterationStart) / 1000000D;
        System.out.printf("%d\t%.3f\n", 3, tookMs);

        final FormIndex indexToDelete2 = new FormIndex(0, 497, mainInstance.getRoot().getChildAt(4).getRef());
        currentIterationStart = System.nanoTime();
        formDef.deleteRepeat(indexToDelete2);
        tookMs = (System.nanoTime() - currentIterationStart) / 1000000D;
        System.out.printf("%d\t%.3f\n", 497, tookMs);
```
Results:
```
3	787.517
497	1.992
```

So I think what you meant is that things speed up as more repeats are deleted.

And what about adding? `createNewRepeat` runs very fast. Does it not result in `sum` being evaluated?

mdudzinski commented:
@lognaturel It does but only once for each new repeat. So evaluating calculations which are dependent on a repeat group and target nodes outside of the repeat group should be improved. My initial idea is to run them once after all "inner" calculations are evaluated for all following repeat groups. I can raise an issue for this with description of my initial idea once this one is merged.

lognaturel commented:
Thank you, that makes sense. Please do file that issue!

mdudzinski commented:
@ln Done #229 !
